pkg/controllers/metrics/node/counter.go

```diff
@@ -0,0 +1,243 @@
+	nodeLabelZone         = corev1.LabelTopologyZone
+
+	nodeConditionTypeReady = corev1.NodeReady
+)
+
+type (
+	nodeListConsumerFunc = func([]corev1.Node) error
+	consumeNodesWithFunc = func(client.MatchingLabels, nodeListConsumerFunc) error
+)
+
+var (
+	nodeLabelProvisioner = karpenterv1.ProvisionerNameLabelKey
+
+	knownValuesForNodeLabels = karpenterv1.WellKnownLabels
+
+	nodeCountByProvisioner = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: metricNamespace,
+			Subsystem: metricSubsystem,
+			Name:      "node_count",
+			Help:      "Total node count by provisioner.",
+		},
+		[]string{
+			metricLabelProvisioner,
+		},
+	)
+
+	readyNodeCountByProvisionerZone = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: metricNamespace,
+			Subsystem: metricSubsystem,
+			Name:      "ready_node_count",
+			Help:      "Count of nodes that are ready by provisioner and zone.",
+		},
+		[]string{
+			metricLabelProvisioner,
+			metricLabelZone,
+		},
+	)
+
+	readyNodeCountByArchProvisionerZone = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: metricNamespace,
+			Subsystem: metricSubsystem,
+			Name:      "ready_node_arch_count",
+			Help:      "Count of nodes that are ready by architecture, provisioner, and zone.",
+		},
+		[]string{
+			metricLabelArch,
+			metricLabelProvisioner,
+			metricLabelZone,
+		},
+	)
+
+	readyNodeCountByInstancetypeProvisionerZone = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: metricNamespace,
+			Subsystem: metricSubsystem,
+			Name:      "ready_node_instancetype_count",
+			Help:      "Count of nodes that are ready by instance type, provisioner, and zone.",
+		},
+		[]string{
+			metricLabelInstanceType,
+			metricLabelProvisioner,
+			metricLabelZone,
+		},
+	)
+
+	readyNodeCountByOsProvisionerZone = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: metricNamespace,
+			Subsystem: metricSubsystem,
+			Name:      "ready_node_os_count",
+			Help:      "Count of nodes that are ready by operating system, provisioner, and zone.",
+		},
+		[]string{
+			metricLabelOs,
+			metricLabelProvisioner,
+			metricLabelZone,
+		},
+	)
+)
+
+func init() {
+	crmetrics.Registry.MustRegister(nodeCountByProvisioner)
+	crmetrics.Registry.MustRegister(readyNodeCountByProvisionerZone)
+	crmetrics.Registry.MustRegister(readyNodeCountByArchProvisionerZone)
+	crmetrics.Registry.MustRegister(readyNodeCountByInstancetypeProvisionerZone)
+	crmetrics.Registry.MustRegister(readyNodeCountByOsProvisionerZone)
+}
+
+func publishNodeCountsForProvisioner(provisioner string, consumeNodesWith consumeNodesWithFunc) error {
+	archValues := knownValuesForNodeLabels[nodeLabelArch]
+	instanceTypeValues := knownValuesForNodeLabels[nodeLabelInstanceType]
+	osValues := knownValuesForNodeLabels[nodeLabelOs]
+	zoneValues := knownValuesForNodeLabels[nodeLabelZone]
+
+	errors := make([]error, 0, len(archValues)*len(instanceTypeValues)*len(osValues)*len(zoneValues))
```

Comment:
I'm wondering if this optimization is necessary? Do we anticipate this slice to get very large?

Reply:
> I'm wondering if this optimization is necessary?

In a test cluster, the total length was >1500 (`len(instanceTypeValues) > 500` and `len(zoneValues) == 3`).

> Do we anticipate this slice to get very large?

The current implementation appends an `error` or `nil` for each metric published.